rs/backend/src/tvl.rs

```diff
@@ -0,0 +1,34 @@
+use crate::{canisters::governance, state::STATE};
+
+pub mod state;
+
+// TODO(NNS1-3281): Remove #[allow(unused)].
+#[allow(unused)]
+pub async fn update_locked_icp_e8s() {
+    let metrics_result = governance::get_metrics().await;
+    STATE.with(|s| {
+        match metrics_result {
+            Ok(Ok(metrics)) => {
+                s.tvl_state.borrow_mut().total_locked_icp_e8s = metrics.total_locked_e8s;
+                ic_cdk::println!("Updated total_locked_icp_e8s for TVL to {}", metrics.total_locked_e8s);
+            }
+            Ok(Err(err)) => {
+                ic_cdk::println!(
+                    "Keeping total_locked_icp_e8s for TVL at {} because of response error: {}",
+                    s.tvl_state.borrow().total_locked_icp_e8s,
+                    err
+                );
+            }
+            Err(err) => {
+                ic_cdk::println!(
+                    "Keeping total_locked_icp_e8s for TVL at {} because of call error: {}",
+                    s.tvl_state.borrow().total_locked_icp_e8s,
+                    err
+                );
+            }
+        };
+    });
```

Comment:
```suggestion
    // Fetch.
    let metrics_result = governance::get_metrics().await;

    // Unpack.
    let metrics_result = match metrics_result {
        Ok(Ok(ok)) => Ok(ok),
        // Else, log and return.
        fail => {
            log(fail);
            return;
        }
    };

    // Store.
    STATE.with(|state| {
        state.tvl_state.borrow_mut().total_locked_icp_e8s = metrics.total_locked_e8s;
    });

    // Conclude.
    println!("Yay! ...");
```

The important thing here is to be really clear what we are using STATE for. If you put the match inside, then, it's harder to see that we only _actually_ use it in the happy case. My alternative makes it super clear when STATE is actually used, and how, because it really minimizes how much code is within `with`.

Reply:
We use it in all 3 cases.
In the error cases it's used to log the current value.
I first had 3 blocks of `STATE.with` but decided to simply it to 1 outer block.

Reply:
NCR

Ah. But logging is an incidental use, not an essential. You could have done

```
// Read current value.
STATE.with(|state| state.borrow().total_locked_icp_e8s)
```

as part of

```
log(fail)
```

I really do not believe in having one giant expression that does everything. Most procedures are pretty "linear"; ofc, they have some branching, but mostly, the branches are little twigs for handling sad along the way. Here, the linear structure is described in the comments that I suggested:

1. Fetch.
2. Unpack response.
3. Store.
4. Report summary.

This really gets lost when you try to "do everything in one big expression".

Reply:
Does this look like what you have in mind?
https://github.com/dfinity/nns-dapp/compare/main...kloet/tvl-match

I agree it makes the main flow more readable. But one thing I don't like about it is that it's matching on the same thing twice and as a result I need to add ` _ => unreachable!(),` in the second `match` so it's no longer "safe".
Do you have a solution for that?